File: ios/Ubapp/Games/Werewolf/WerewolfServer.swift
import Foundation

/// Wraps [HostServer] with Werewolf-specific routing. Mirrors
/// lib/games/werewolf/werewolf_server.dart.
final class WerewolfServer {
    let engine = WerewolfEngine()
    static let hostId = "host"
    let hostName: String

    private let server: HostServer
    private var guestToPlayer: [GuestId: String] = [:]
    private var playerToGuest: [String: GuestId] = [:]

    var onStateChange: (() -> Void)?

    init(server: HostServer? = nil, hostName: String = "Host") {
        self.server = server ?? HostServer(html: HostServer.htmlResource(named: "werewolf_browser"))
        self.hostName = hostName
        self.server.onMessage = { [weak self] id, raw in self?.onMessage(from: id, raw: raw) }
        self.server.onLeave = { [weak self] id in self?.onLeave(id) }
    }

    @discardableResult
    func start() throws -> URL? {
        engine.addPlayer(id: Self.hostId, name: hostName, isHost: true)
        let url = try server.start()
        let local = server.attachLocalGuest()
        guestToPlayer[local] = Self.hostId
        playerToGuest[Self.hostId] = local
        emit()
        return url
    }

    /// In-process pipe for the host's own player view.
    @MainActor
    func makeLoopback() -> LoopbackGuest { LoopbackGuest(server: server) }
    func stop() { server.stop() }
    var guestCount: Int { server.guestCount }

    // MARK: Host-side actions
    func hostSetOptions(_ o: WerewolfOptions) {
        engine.setOptions(o)
        broadcastOptions(); emit()
    }
    func hostStart() {
        guard engine.canStart else { return }
        engine.start()
        broadcastPhase()
        sendRolesPrivately()
        emit()
    }
    func hostNightAction(targetId: String) { applyNightAction(playerId: Self.hostId, targetId: targetId) }
    func hostDayVote(targetId: String?) { applyDayVote(playerId: Self.hostId, targetId: targetId) }
    func hostHunterShot(targetId: String) { applyHunterShot(playerId: Self.hostId, targetId: targetId) }
    func hostCallTutorialVote() { openTutorialVote() }
    func hostTutorialVote(_ yes: Bool) { submitTutorialVote(voterId: Self.hostId, yes: yes) }
    func hostDismissTutorial() {
        engine.tutorialVote.markShown()
        broadcastTutorialState()
        emit()
    }
    func advanceFromReveal() {
        engine.advanceToDayVote()
        if engine.phase == .gameOver { broadcastGameOver() } else { broadcastPhase() }
        emit()
    }

    // MARK: Inbound
    private func onMessage(from guest: GuestId, raw: String) {
        guard let data = raw.data(using: .utf8),
              let j = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = j["type"] as? String else { return }
        switch type {
        case "join": handleJoin(guest, json: j)
        case "night_action":
            if let pid = guestToPlayer[guest], let t = j["targetId"] as? String {
                applyNightAction(playerId: pid, targetId: t)
            }
        case "vote":
            if let pid = guestToPlayer[guest] {
                applyDayVote(playerId: pid, targetId: j["targetId"] as? String)
            }
        case "hunter_shot":
            if let pid = guestToPlayer[guest], let t = j["targetId"] as? String {
                applyHunterShot(playerId: pid, targetId: t)
            }
        case "set_options":
            // Only the host (not over WebSocket) may mutate options.
            break
        case "call_tutorial_vote": openTutorialVote()
        case "tutorial_vote":
            if let pid = guestToPlayer[guest], let yes = j["yes"] as? Bool {
                submitTutorialVote(voterId: pid, yes: yes)
            }
        default: break
        }
    }

    private func onLeave(_ guest: GuestId) {
        guard let pid = guestToPlayer.removeValue(forKey: guest) else { return }
        playerToGuest[pid] = nil
        engine.removePlayer(pid)
        engine.tutorialVote.removeVoter(pid)
        broadcastLobby()
        if engine.tutorialVote.isOpen || engine.tutorialVote.hasResult {
            broadcastTutorialState()
        }
        emit()
    }

    private func handleJoin(_ guest: GuestId, json: [String: Any]) {
        if engine.phase != .lobby {
            send(guest, ["type": "error", "message": "Game already started"])
            return
        }
        let name = (json["name"] as? String)?.trimmingCharacters(in: .whitespaces) ?? ""
        guard !name.isEmpty else { return }
        let pid = "g\(guestToPlayer.count + 1)"
        engine.addPlayer(id: pid, name: name)
        guestToPlayer[guest] = pid
        playerToGuest[pid] = guest
        send(guest, ["type": "welcome", "yourId": pid, "yourName": name, "game": "werewolf"])
        broadcastLobby()
        broadcastOptions()
        broadcastTutorialState()
        emit()
    }

    private func broadcastOptions() {
        let o = engine.options
        broadcast([
            "type": "options",
            "wolfCount": o.wolfCount as Any,
            "seerEnabled": o.seerEnabled,
            "hunterEnabled": o.hunterEnabled,
            "maxWolfCount": engine.maxWolfCount,
        ])
    }

    private func applyNightAction(playerId: String, targetId: String) {
        guard let p = engine.players[playerId], p.alive else { return }
        var ready = false
        if p.role == .werewolf {
            ready = engine.submitWolfVote(voterId: playerId, targetId: targetId)
        } else if p.role == .seer {
            ready = engine.submitSeerTarget(seerId: playerId, targetId: targetId)
        }
        emit()
        if ready {
            engine.resolveNight()
            sendSeerResultPrivately()
            broadcastNightResult()
            if engine.phase == .gameOver { broadcastGameOver() }
            else if engine.phase == .hunterShot { broadcastHunterPrompt() }
            emit()
        }
    }

    private func applyDayVote(playerId: String, targetId: String?) {
        let ready = engine.submitDayVote(voterId: playerId, targetId: targetId)
        broadcastVoteUpdate()
        emit()
        if ready {
            engine.resolveDay()
            broadcastDayResult()
            if engine.phase == .gameOver { broadcastGameOver() }
            else if engine.phase == .hunterShot { broadcastHunterPrompt() }
            else { broadcastPhase() }
            emit()
        }
    }

    private func applyHunterShot(playerId: String, targetId: String) {
        let ok = engine.submitHunterShot(hunterId: playerId, targetId: targetId)
        if !ok { return }
        broadcastHunterShotResult()
        if engine.phase == .gameOver { broadcastGameOver() }
        else if engine.phase == .hunterShot { broadcastHunterPrompt() }
        else { broadcastPhase() }
        emit()
    }

    // MARK: Outbound
    private func broadcastLobby() {
        broadcast([
            "type": "lobby",
            "players": engine.players.values.map { ["id": $0.id, "name": $0.name, "isHost": $0.isHost] },
            "canStart": engine.canStart,
        ])
    }

    private func openTutorialVote() {
        guard engine.phase == .lobby, !engine.tutorialVote.isOpen, !engine.tutorialVote.tutorialShown else { return }
        engine.tutorialVote.open(eligibleIds: engine.players.keys)
        broadcastTutorialState(); emit()
    }
    private func submitTutorialVote(voterId: String, yes: Bool) {
        guard engine.tutorialVote.isOpen else { return }
        engine.tutorialVote.submit(voterId: voterId, yes: yes)
        broadcastTutorialState(); emit()
    }
    private func broadcastTutorialState() {
        let v = engine.tutorialVote
        var payload: [String: Any] = [
            "type": "tutorial_vote_state",
            "isOpen": v.isOpen, "yesCount": v.yesCount, "noCount": v.noCount,
            "eligibleCount": v.eligibleCount, "result": v.result as Any, "tutorialShown": v.tutorialShown,
        ]
        if v.result == true && !v.tutorialShown {
            payload["title"] = GameTutorials.werewolf.title
            payload["sections"] = GameTutorials.werewolf.sectionsJSON()
            payload["menuSections"] = GameTutorials.werewolf.browserMenuSectionsJSON()
        }
        broadcast(payload)
    }

    private func sendRolesPrivately() {
        let wolfIds = engine.players.values.filter { $0.role == .werewolf }.map(\.id)
        for p in engine.players.values {
            var payload: [String: Any] = ["type": "role", "role": p.role!.rawValue]
            if p.role == .werewolf { payload["wolfIds"] = wolfIds }
            if let guest = playerToGuest[p.id] { send(guest, payload) }
        }
    }

    private func sendSeerResultPrivately() {
        guard let r = engine.lastSeerResult, let guest = playerToGuest[r.seerId] else { return }
        send(guest, ["type": "seer_result", "targetId": r.targetId, "isWerewolf": r.isWerewolf])
    }

    private func broadcastPhase() {
        broadcast([
            "type": "phase",
            "phase": phaseName(engine.phase),
            "day": engine.day,
            "alive": engine.alive.map(publicPlayer),
            "dead": engine.dead.map(publicPlayer),
        ])
    }
    private func broadcastVoteUpdate() {
        var votes: [String: String] = [:]
        for (k, v) in engine.dayVotes { votes[k] = v ?? "" }
        broadcast(["type": "vote_update", "votes": votes])
    }
    private func broadcastNightResult() {
        let n = engine.lastNight!
        broadcast([
            "type": "phase",
            "phase": phaseName(engine.phase),
            "day": engine.day,
            "alive": engine.alive.map(publicPlayer),
            "dead": engine.dead.map(publicPlayer),
            "killedId": n.killedId as Any,
        ])
    }
    private func broadcastDayResult() {
        let d = engine.lastDay!
        let role: Any = d.eliminatedId.flatMap { engine.players[$0]?.role?.rawValue } ?? NSNull()
        broadcast([
            "type": "day_result",
            "eliminatedId": d.eliminatedId as Any,
            "tally": d.tally,
            "alive": engine.alive.map(publicPlayer),
            "dead": engine.dead.map(publicPlayer),
            "eliminatedRole": role,
        ])
    }
    private func broadcastHunterPrompt() {
        broadcast([
            "type": "hunter_prompt",
            "hunterId": engine.pendingHunterShooter as Any,
            "alive": engine.alive.map(publicPlayer),
            "dead": engine.dead.map(publicPlayer),
        ])
    }
    private func broadcastHunterShotResult() {
        guard let last = engine.hunterShotsThisRound.last else { return }
        broadcast([
            "type": "hunter_shot_result",
            "hunterId": last.hunterId,
            "targetId": last.targetId,
            "targetRole": engine.players[last.targetId]!.role!.rawValue,
            "alive": engine.alive.map(publicPlayer),
            "dead": engine.dead.map(publicPlayer),
        ])
    }
    private func broadcastGameOver() {
        var roles: [String: String] = [:]
        for p in engine.players.values { roles[p.id] = p.role!.rawValue }
        broadcast([
            "type": "game_over",
            "winner": engine.winner == .town ? "town" : "werewolves",
            "roles": roles,
        ])
    }

    private func publicPlayer(_ p: WerewolfPlayer) -> [String: Any] {
        ["id": p.id, "name": p.name, "alive": p.alive]
    }

    private func phaseName(_ p: WerewolfPhase) -> String {
        switch p {
        case .lobby: "lobby"; case .night: "night"
        case .dayReveal: "dayReveal"; case .dayVote: "dayVote"
        case .hunterShot: "hunterShot"; case .gameOver: "gameOver"
        }
    }

    private func emit() { onStateChange?() }
    private func broadcast(_ payload: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let s = String(data: data, encoding: .utf8) else { return }
        server.broadcast(s)
    }
    private func send(_ guest: GuestId, _ payload: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let s = String(data: data, encoding: .utf8) else { return }
        server.send(to: guest, s)
    }
}

File: android/app/src/main/java/com/example/jamboree/games/werewolf/WerewolfServer.kt
package com.example.jamboree.games.werewolf

import android.content.Context
import com.example.jamboree.join.LoopbackGuest
import com.example.jamboree.social.GuestId
import com.example.jamboree.social.HostServer
import com.example.jamboree.stats.StatsStore
import com.example.jamboree.tutorials.GameTutorials
import org.json.JSONArray
import org.json.JSONObject

/** Wraps [HostServer] with Werewolf-specific routing. */
class WerewolfServer(context: Context, val hostName: String = "Host") {
    var engine = WerewolfEngine(); private set
    private val server = HostServer(html = HostServer.htmlAsset(context, "werewolf_browser.html"), ctx = context)
    private val appCtx = context.applicationContext
    private val guestToPlayer = HashMap<GuestId, String>()
    private val playerToGuest = HashMap<String, GuestId>()
    var onStateChange: (() -> Unit)? = null
    private var statRecorded = false

    init { server.onMessage = ::onMessage; server.onLeave = ::onLeave }
    companion object { const val HOST_ID = "host" }

    fun start(): String? {
        engine.addPlayer(HOST_ID, hostName, isHost = true)
        val u = server.startServer()
        val local = server.attachLocalGuest()
        guestToPlayer[local] = HOST_ID
        playerToGuest[HOST_ID] = local
        emit()
        return u
    }

    /** In-process pipe for the host's own player screen. */
    fun makeLoopback(): LoopbackGuest = LoopbackGuest(server)

    fun stop() { server.stopServer(); resetState() }

    /** Clear all per-session state so the next time the host starts
     *  hosting they get a fresh screen — empty lobby, tutorial vote
     *  available again. */
    private fun resetState() {
        engine = WerewolfEngine()
        guestToPlayer.clear(); playerToGuest.clear()
        statRecorded = false
        emit()
    }
    val guestCount: Int get() = server.guestCount

    fun hostSetOptions(o: WerewolfOptions) {
        engine.setOptions(o); broadcastOptions(); emit()
    }
    fun hostStart() { if (!engine.canStart) return; engine.start(); broadcastPhase(); sendRolesPrivately(); emit() }
    fun hostNightAction(targetId: String) = applyNightAction(HOST_ID, targetId)
    fun hostDayVote(targetId: String?) = applyDayVote(HOST_ID, targetId)
    fun hostHunterShot(targetId: String) = applyHunterShot(HOST_ID, targetId)
    fun hostCallTutorialVote() = openTutorialVote()
    fun hostTutorialVote(yes: Boolean) = submitTutorialVote(HOST_ID, yes)
    fun hostDismissTutorial() { engine.tutorialVote.markShown(); broadcastTutorialState(); emit() }
    fun advanceFromReveal() {
        engine.advanceToDayVote()
        if (engine.phase == WerewolfPhase.GAME_OVER) broadcastGameOver() else broadcastPhase()
        emit()
    }

    private fun onMessage(guest: GuestId, raw: String) {
        val j = runCatching { JSONObject(raw) }.getOrNull() ?: return
        when (j.optString("type")) {
            "join" -> handleJoin(guest, j)
            "night_action" -> guestToPlayer[guest]?.let { applyNightAction(it, j.getString("targetId")) }
            "vote" -> guestToPlayer[guest]?.let {
                applyDayVote(it, if (j.isNull("targetId")) null else j.getString("targetId"))
            }
            "hunter_shot" -> guestToPlayer[guest]?.let { applyHunterShot(it, j.getString("targetId")) }
            // Only the host (not over WebSocket) may mutate options.
            "set_options" -> Unit
            "call_tutorial_vote" -> guestToPlayer[guest]?.let { openTutorialVote() }
            "tutorial_vote" -> guestToPlayer[guest]?.let { submitTutorialVote(it, j.getBoolean("yes")) }
        }
    }

    private fun onLeave(guest: GuestId) {
        val pid = guestToPlayer.remove(guest) ?: return
        playerToGuest.remove(pid)
        engine.removePlayer(pid); engine.tutorialVote.removeVoter(pid)
        broadcastLobby()
        if (engine.tutorialVote.isOpen || engine.tutorialVote.hasResult) broadcastTutorialState()
        emit()
    }

    private fun handleJoin(guest: GuestId, j: JSONObject) {
        if (engine.phase != WerewolfPhase.LOBBY) {
            send(guest, JSONObject().put("type", "error").put("message", "Game already started")); return
        }
        val name = j.optString("name").trim().take(24); if (name.isEmpty()) return
        val pid = "g${guestToPlayer.size + 1}"
        engine.addPlayer(pid, name)
        guestToPlayer[guest] = pid; playerToGuest[pid] = guest
        send(guest, JSONObject().put("type", "welcome").put("yourId", pid).put("yourName", name).put("game", "werewolf"))
        broadcastLobby(); broadcastOptions(); broadcastTutorialState(); emit()
    }

    private fun broadcastOptions() {
        val o = engine.options
        broadcast(JSONObject()
            .put("type", "options")
            .put("wolfCount", o.wolfCount ?: JSONObject.NULL)
            .put("seerEnabled", o.seerEnabled)
            .put("hunterEnabled", o.hunterEnabled)
            .put("maxWolfCount", engine.maxWolfCount))
    }

    private fun applyNightAction(playerId: String, targetId: String) {
        val p = engine.players[playerId] ?: return
        if (!p.alive) return
        val ready = when (p.role) {
            WerewolfRole.WEREWOLF -> engine.submitWolfVote(playerId, targetId)
            WerewolfRole.SEER -> engine.submitSeerTarget(playerId, targetId)
            else -> false
        }
        emit()
        if (ready) {
            engine.resolveNight(); sendSeerResultPrivately(); broadcastNightResult()
            if (engine.phase == WerewolfPhase.GAME_OVER) broadcastGameOver()
            else if (engine.phase == WerewolfPhase.HUNTER_SHOT) broadcastHunterPrompt()
            emit()
        }
    }

    private fun applyDayVote(playerId: String, targetId: String?) {
        val ready = engine.submitDayVote(playerId, targetId); broadcastVoteUpdate(); emit()
        if (ready) {
            engine.resolveDay(); broadcastDayResult()
            if (engine.phase == WerewolfPhase.GAME_OVER) broadcastGameOver()
            else if (engine.phase == WerewolfPhase.HUNTER_SHOT) broadcastHunterPrompt()
            else broadcastPhase()
            emit()
        }
    }

    private fun applyHunterShot(playerId: String, targetId: String) {
        if (!engine.submitHunterShot(playerId, targetId)) return
        broadcastHunterShotResult()
        if (engine.phase == WerewolfPhase.GAME_OVER) broadcastGameOver()
        else if (engine.phase == WerewolfPhase.HUNTER_SHOT) broadcastHunterPrompt()
        else broadcastPhase()
        emit()
    }

    private fun broadcastLobby() {
        val arr = JSONArray()
        for (p in engine.players.values) arr.put(JSONObject().put("id", p.id).put("name", p.name).put("isHost", p.isHost))
        broadcast(JSONObject().put("type", "lobby").put("players", arr).put("canStart", engine.canStart))
    }
    private fun sendRolesPrivately() {
        val wolfIds = JSONArray(engine.players.values.filter { it.role == WerewolfRole.WEREWOLF }.map { it.id })
        for (p in engine.players.values) {
            val payload = JSONObject().put("type", "role").put("role", roleName(p.role!!))
            if (p.role == WerewolfRole.WEREWOLF) payload.put("wolfIds", wolfIds)
            playerToGuest[p.id]?.let { send(it, payload) }
        }
    }
    private fun sendSeerResultPrivately() {
        val r = engine.lastSeerResult ?: return
        val guest = playerToGuest[r.seerId] ?: return
        send(guest, JSONObject().put("type", "seer_result").put("targetId", r.targetId).put("isWerewolf", r.isWerewolf))
    }
    private fun broadcastPhase() {
        broadcast(JSONObject()
            .put("type", "phase").put("phase", phaseJson(engine.phase)).put("day", engine.day)
            .put("alive", publicArr(engine.alive)).put("dead", publicArr(engine.dead)))
    }
    private fun broadcastVoteUpdate() {
        val votes = JSONObject(); for ((k, v) in engine.dayVotes) votes.put(k, v ?: "")
        broadcast(JSONObject().put("type", "vote_update").put("votes", votes))
    }
    private fun broadcastNightResult() {
        val n = engine.lastNight!!
        broadcast(JSONObject()
            .put("type", "phase").put("phase", phaseJson(engine.phase)).put("day", engine.day)
            .put("alive", publicArr(engine.alive)).put("dead", publicArr(engine.dead))
            .put("killedId", n.killedId ?: JSONObject.NULL))
    }
    private fun broadcastDayResult() {
        val d = engine.lastDay!!
        val role = d.eliminatedId?.let { engine.players[it]?.role?.let { r -> roleName(r) } }
        broadcast(JSONObject()
            .put("type", "day_result")
            .put("eliminatedId", d.eliminatedId ?: JSONObject.NULL)
            .put("tally", JSONObject(d.tally as Map<*, *>))
            .put("alive", publicArr(engine.alive)).put("dead", publicArr(engine.dead))
            .put("eliminatedRole", role ?: JSONObject.NULL))
    }
    private fun broadcastHunterPrompt() {
        broadcast(JSONObject()
            .put("type", "hunter_prompt")
            .put("hunterId", engine.pendingHunterShooter ?: JSONObject.NULL)
            .put("alive", publicArr(engine.alive)).put("dead", publicArr(engine.dead)))
    }
    private fun broadcastHunterShotResult() {
        val last = engine.hunterShotsThisRound.lastOrNull() ?: return
        broadcast(JSONObject()
            .put("type", "hunter_shot_result")
            .put("hunterId", last.hunterId).put("targetId", last.targetId)
            .put("targetRole", roleName(engine.players[last.targetId]!!.role!!))
            .put("alive", publicArr(engine.alive)).put("dead", publicArr(engine.dead)))
    }
    private fun broadcastGameOver() {
        if (!statRecorded) {
            statRecorded = true
            StatsStore.record(
                appCtx, "werewolf",
                engine.players.values.map { it.name },
                if (engine.winner == WerewolfWinner.TOWN) "town" else "werewolves",
            )
        }
        val roles = JSONObject()
        for (p in engine.players.values) roles.put(p.id, roleName(p.role!!))
        broadcast(JSONObject().put("type", "game_over")
            .put("winner", if (engine.winner == WerewolfWinner.TOWN) "town" else "werewolves")
            .put("roles", roles))
    }

    private fun openTutorialVote() {
        if (engine.phase != WerewolfPhase.LOBBY) return
        if (engine.tutorialVote.isOpen || engine.tutorialVote.tutorialShown) return
        engine.tutorialVote.open(engine.players.keys); broadcastTutorialState(); emit()
    }
    private fun submitTutorialVote(voterId: String, yes: Boolean) {
        if (!engine.tutorialVote.isOpen) return
        engine.tutorialVote.submit(voterId, yes); broadcastTutorialState(); emit()
    }
    private fun broadcastTutorialState() {
        val v = engine.tutorialVote
        val p = JSONObject().put("type", "tutorial_vote_state")
            .put("isOpen", v.isOpen).put("yesCount", v.yesCount).put("noCount", v.noCount)
            .put("eligibleCount", v.eligibleCount).put("result", v.result ?: JSONObject.NULL)
            .put("tutorialShown", v.tutorialShown)
        if (v.result == true && !v.tutorialShown) {
            p.put("title", GameTutorials.werewolf.title)
            p.put("sections", JSONArray(GameTutorials.werewolf.sectionsJson().map { JSONObject(it as Map<*, *>) }))
            p.put("menuSections", JSONArray(GameTutorials.werewolf.browserMenuSectionsJson().map { JSONObject(it as Map<*, *>) }))
        }
        broadcast(p)
    }

    private fun phaseJson(p: WerewolfPhase) = when (p) {
        WerewolfPhase.LOBBY -> "lobby"; WerewolfPhase.NIGHT -> "night"
        WerewolfPhase.DAY_REVEAL -> "dayReveal"; WerewolfPhase.DAY_VOTE -> "dayVote"
        WerewolfPhase.HUNTER_SHOT -> "hunterShot"; WerewolfPhase.GAME_OVER -> "gameOver"
    }
    private fun roleName(r: WerewolfRole) = r.name.lowercase()

    private fun publicArr(ps: List<WerewolfPlayer>): JSONArray {
        val a = JSONArray()
        for (p in ps) a.put(JSONObject().put("id", p.id).put("name", p.name).put("alive", p.alive))
        return a
    }
    private fun emit() { onStateChange?.invoke() }
    private fun broadcast(o: JSONObject) = server.broadcast(o.toString())
    private fun send(g: GuestId, o: JSONObject) = server.send(g, o.toString())
}
